### ISSUE TITLE
Visual Studio crashed in IncrementalAnalyzerProcessor

### ISSUE THREAD

**Issue by nigurr** [dotnet/roslyn#26171]
**Version Used**: 15.7 Preview 3

**Steps to Reproduce**:

1. Opening any project/solution crashes VS

Stack trace:
> Application: devenv.exe
> Framework Version: v4.0.30319
> Description: The application requested process termination through System.Environment.FailFast(string message).
> Message: System.ObjectDisposedException: Cannot access a disposed object.
> Object name: 'StreamJsonRpc.JsonRpc'.
>    at Microsoft.Verify.NotDisposed(IDisposableObservable disposedValue, String message)
>    at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__67`1.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
>    at Microsoft.VisualStudio.LanguageServices.Remote.ServiceHubRemoteHostClient.<CreateWorkerAsync>d__11.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.VisualStudio.LanguageServices.Remote.ServiceHubRemoteHostClient.Connections.<RetryRemoteCallAsync>d__0`2.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.VisualStudio.LanguageServices.Remote.ServiceHubRemoteHostClient.<CreateAsync>d__10.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.VisualStudio.LanguageServices.Remote.RemoteHostClientFactory.<CreateAsync>d__0.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.VisualStudio.LanguageServices.Remote.RemoteHostClientServiceFactory.RemoteHostClientService.<EnableAsync>d__18.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.CodeAnalysis.Remote.RemoteHostClientExtensions.<SynchronizePrimaryWorkspaceAsync>d__23.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService.WorkCoordinator.IncrementalAnalyzerProcessor.NormalPriorityProcessor.<ResetStatesAsync>d__37.MoveNext()
> Stack:
>    at System.Environment.FailFast(System.String, System.Exception)
>    at Microsoft.CodeAnalysis.FailFast.OnFatalException(System.Exception)
>    at Microsoft.CodeAnalysis.ErrorReporting.FatalError.Report(System.Exception, System.Action`1<System.Exception>)
>    at Microsoft.CodeAnalysis.ErrorReporting.FatalError.ReportUnlessCanceled(System.Exception)
>    at Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService+WorkCoordinator+IncrementalAnalyzerProcessor+NormalPriorityProcessor+<ResetStatesAsync>d__37.MoveNext()
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
>    at Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService+WorkCoordinator+IncrementalAnalyzerProcessor+NormalPriorityProcessor+<ResetStatesAsync>d__37.MoveNext()
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService+WorkCoordinator+IncrementalAnalyzerProcessor+NormalPriorityProcessor+<ResetStatesAsync>d__37, Microsoft.CodeAnalysis.Features, Version=2.8.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]](<ResetStatesAsync>d__37 ByRef)
>    at Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService+WorkCoordinator+IncrementalAnalyzerProcessor+NormalPriorityProcessor.ResetStatesAsync()
>    at Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService+WorkCoordinator+IncrementalAnalyzerProcessor+NormalPriorityProcessor+<ExecuteAsync>d__17.MoveNext()
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService+WorkCoordinator+IncrementalAnalyzerProcessor+NormalPriorityProcessor+<ExecuteAsync>d__17, Microsoft.CodeAnalysis.Features, Version=2.8.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]](<ExecuteAsync>d__17 ByRef)
>    at Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService+WorkCoordinator+IncrementalAnalyzerProcessor+NormalPriorityProcessor.ExecuteAsync()
>    at Microsoft.CodeAnalysis.SolutionCrawler.IdleProcessor+<ProcessAsync>d__12.MoveNext()
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(System.Object)
>    at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
>    at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
>    at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action, Boolean, System.Threading.Tasks.Task ByRef)
>    at System.Threading.Tasks.Task.FinishContinuations()
>    at System.Threading.Tasks.Task.FinishStageThree()
>    at System.Threading.Tasks.Task`1[[System.Threading.Tasks.VoidTaskResult, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].TrySetResult(System.Threading.Tasks.VoidTaskResult)
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].SetResult(System.Threading.Tasks.VoidTaskResult)
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].SetResult(System.Threading.Tasks.Task`1<System.Threading.Tasks.VoidTaskResult>)
>    at Microsoft.CodeAnalysis.SolutionCrawler.IdleProcessor+<WaitForIdleAsync>d__11.MoveNext()
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(System.Object)
>    at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
>    at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
>    at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action, Boolean, System.Threading.Tasks.Task ByRef)
>    at System.Threading.Tasks.Task.FinishContinuations()
>    at System.Threading.Tasks.Task.FinishStageThree()
>    at System.Threading.Tasks.Task`1[[System.Threading.Tasks.VoidTaskResult, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].TrySetResult(System.Threading.Tasks.VoidTaskResult)
>    at System.Threading.Tasks.Task+DelayPromise.Complete()
>    at System.Threading.Tasks.Task+<>c.<Delay>b__274_1(System.Object)
>    at System.Threading.TimerQueueTimer.CallCallbackInContext(System.Object)
>    at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
>    at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
>    at System.Threading.TimerQueueTimer.CallCallback()
>    at System.Threading.TimerQueueTimer.Fire()
>    at System.Threading.TimerQueue.FireNextTimers()
>    at System.Threading.TimerQueue.AppDomainTimerCallback()
> 

I can share the repro machine/dump if required.

**Comment by heejaechang** [comment]
Hi Thank you for reporting the issue?

can you share logs in %temp%\servicehub\logs? looks like some setup is messed up. the callstack shows VS is trying to start out of proc process and that failed to be started.

**Comment by heejaechang** [comment]
@jimmy9988 En-Shuo, have you reported this issue to FSharp team? it looks like somehow FSharp setup puts this in devenv.exe.config

```
			<dependentAssembly>
				<assemblyIdentity name="Newtonsoft.Json" publicKeyToken="30ad4fe6b2a6aeed" culture="neutral"/>
				<codeBase version="10.0.0.0" href="C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\FSharp\Newtonsoft.Json.dll"/>
			</dependentAssembly>
```

but some VS drop doesn't have newtonsoft.json.dll in FSharp directory. causing everything that uses newtonsoft.json.dll in VS to fail.

another thing is that, it points to version 10.0.0.0 but config has binding redirect to 9.0.0.0

```
<dependentAssembly>
				<assemblyIdentity name="Newtonsoft.Json" publicKeyToken="30ad4fe6b2a6aeed" culture="neutral"/>
				<bindingRedirect oldVersion="1.0.0.0-9.0.0.0" newVersion="9.0.0.0"/>
			</dependentAssembly>
```
making 2 different versions of newtonsoft.json.dll to be used.

are these known issue?

tagging @jinujoseph @KevinRansom @brettfo 

...

also, one of reliability bug @ivanbasov assigned to me (service hub failing to start) seems related to this.

**Comment by heejaechang** [comment]
@nigurr can you check your directory (C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\FSharp\Newtonsoft.Json.dll) to see whether it has newtonsoft.json.dll ?

**Comment by heejaechang** [comment]
here are service hub logs @nigurr shared. basically multiple services fail due to newtonsoft.json.dll issue.

...

04/17/2018 10:26:15 : Error : Error creating Microsoft.VisualStudio.ExtensionManager.ServiceModule.ExtensionManagerServiceModule instance of CLR service module 'extensionManager': System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. HResult='-2146232828' 
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.ServiceHub.HostLib.ServiceManager.CreateAndConfigureService(String name, Type serviceModuleType, WrappedStream stream)
===InnerException===
System.IO.FileNotFoundException: Could not load file or assembly 'Newtonsoft.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The system cannot find the file specified. HResult='-2147024894' 
   at Microsoft.VisualStudio.ExtensionManager.ServiceModule.ExtensionManagerServiceModule..ctor(Stream serviceHubStream, IServiceProvider serviceProvider)

04/17/2018 10:26:16 : Error : Error creating Microsoft.CodeAnalysis.Remote.RemoteHostService instance of CLR service module 'roslynRemoteHost': System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. HResult='-2146232828' 
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.ServiceHub.HostLib.ServiceManager.CreateAndConfigureService(String name, Type serviceModuleType, WrappedStream stream)
===InnerException===
System.IO.FileNotFoundException: Could not load file or assembly 'Newtonsoft.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The system cannot find the file specified. HResult='-2147024894' 
   at Microsoft.CodeAnalysis.Remote.RemoteHostService..ctor(Stream stream, IServiceProvider serviceProvider)
04/17/2018 10:26:16 : Error : Error creating Microsoft.CodeAnalysis.Remote.SnapshotService instance of CLR service module 'roslynSnapshot': System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. HResult='-2146232828' 
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.ServiceHub.HostLib.ServiceManager.CreateAndConfigureService(String name, Type serviceModuleType, WrappedStream stream)
===InnerException===
System.IO.FileNotFoundException: Could not load file or assembly 'Newtonsoft.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The system cannot find the file specified. HResult='-2147024894' 
   at Microsoft.CodeAnalysis.Remote.SnapshotService..ctor(Stream stream, IServiceProvider serviceProvider)

**Comment by nigurr** [comment]
@heejaechang I am using latest preview and Newton.Json doesn't exist in `C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\Common7\IDE\CommonExtensions\Microsoft\FSharp`

**Comment by jinujoseph** [comment]
cc @Pilchie  for info

**Comment by heejaechang** [comment]
@nigurr thank you

ping @KevinRansom @brettfo 

according to @ivanbasov this is high priority
> this is top crash in 15.7. preview 3 across the whole VS. Could you please take a look? It is a new regression started in 15.7

**Comment by heejaechang** [comment]
@jinujoseph regardless this issue, I think we need to clean up newtonsoft.json.dll mess in VS. I am not sure where to start discussion on this.

can we start it somehow?

I have 15.6.4 installed. and these are how many newtonsoft.json.dll installed by VS. most of them exact same copy. I believe only this (yellow highlighted) should be there

...

![image](https://user-images.githubusercontent.com/1333179/38853446-1a88413c-41d2-11e8-8fe4-d36c58443cb7.png)

**Comment by Pilchie** [comment]
@brettfo deliberately removed that copy, so that F# would use the same version as everyone else, but it sounds like a the `devenv.exe.config` didn't get updated correctly?

**Comment by Pilchie** [comment]
Can we get a VSTS bug on this for shiproom tracking ASAP please?

**Comment by brettfo** [comment]
I think the Newtonsoft issue is a red herring.  I've been running 15.7 previews for weeks and none of them have a copy of newtonsoft.json present.

The devenv.exe.config issue certainly needs to be fixed, but that would only cause issues at install time when everything is NGENed.

**Comment by heejaechang** [comment]
I investigated this issue with @nigurr before. once we either copied newtonsoft.json to the FSharp folder or remove newtonsoft.json codebase to the FSharp folder, VS worked as expected.

@jimmy9988 can you copy over newtonsoft.json to the FSharp folder and see whether VS still crash?

**Comment by brettfo** [comment]
I have a new theory as to what happened:

1.  A VS preview was installed with F# support back when F# shipped Newtonsoft.Json.dll and the appropriate entry was added to the hive-local devenv.exe.config.
2.  The previews were upgraded past the point where Newtonsoft.Json.dll was no longer included.
3.  devenv.exe.config was never regenerated.

@jimmy9988 Can you try again with a clean install of VS _and_ after deleting your hives under `%localappdata%\Microsoft\VisualStudio\15.0*` and %appdata%\Microsoft\VisualStudio\15.0*`?  If the problem still persists then I think the next step is to look into how/when devenv.exe.config is regenerated.

**Comment by nigurr** [comment]
@Pilchie I created DevDiv bug internally - 601243
Please update the details accordingly

**Comment by jinujoseph** [comment]
watson vso issue tracking this 600984

**Comment by brettfo** [comment]
Fixed in Microsoft/visualfsharp#4764.  The 15.7 release of VS will contain the fix.
